lib/Service/RowService.php

```diff
@@ -283,11 +284,22 @@ private function enhanceWithViewDefaults(?View $view, RowDataInput $data): RowDa
 	/**
 	 * @throws InternalError
 	 */
-	private function cleanupData(RowDataInput $data, array $columns, ?int $tableId, ?int $viewId): RowDataInput {
+	private function cleanupAndValidateData(RowDataInput $data, array $columns, ?int $tableId, ?int $viewId, ?int $rowId = null): RowDataInput {
 		$out = new RowDataInput();
+		$tableColumns = $this->columnMapper->findAllByTable($tableId);
 		foreach ($data as $entry) {
 			$column = $this->getColumnFromColumnsArray((int)$entry['columnId'], $columns);
 
+			if ($column && $column->getType() === Column::TYPE_TEXT
+				&& $column->getSubtype() ===  Column::SUBTYPE_TEXT_LINE
+				&& $column->getTextUnique()) {
+				$filter = [[['columnId' => $column->getId(), 'operator' => 'is-equal', 'value' => $entry['value']]]];
+				$alreadyExistentRows = $this->row2Mapper->findAll($tableColumns, $columns, $tableId, filter: $filter, userId: $this->userId);
+				if ($alreadyExistentRows && $alreadyExistentRows[0]->getId() !== $rowId) {
+					throw new BadRequestError('Column with id ' . $entry['columnId'] . ' is not unique.');
+				}
+			}
```

Comment:
I would prefer to have the specific logic put elsewhere, somewhere closer to the subtype related code parts. The validation itself is ok, I am afraid we would have a blown up method here if more future types would add validation here. Not a blocker.

Reply:
I like this idea. Moved validation to corresponding "column business" class